### PR TITLE
[flang] allow intrinsic module procedures to be implemented in Fortran

### PR DESCRIPTION
**Pull request by jeanPerier**
Currently, all procedures from intrinsic modules that are not BIND(C) are expected to be intercepted by the compiler in lowering and to have a handler in IntrinsicCall.cpp.

As more "intrinsic" modules are being added (OpenMP, OpenACC, CUF, ...), this requirement is preventing seamless implementation of intrinsic modules in Fortran. Procedures from intrinsic modules are different from generic intrinsics defined in section 16 of the standard. They are declared in Fortran file seating in the intrinsic module directory and inside the compiler they look like regular user call except for the INTRINSIC attribute set on their module. So an easy implementation is just to have the implementation done in Fortran and linked to the runtime without any need for the compiler to necessarily understand and handle these calls in special ways.

This patch splits the lookup and generation part of IntrinsicCall.cpp so that it can be allowed to only intercept calls to procedure from intrinsic module if they have a handler. Otherwise, the assumption is that they should be implemented in Fortran.

Add explicit TODOs handler for the IEEE procedure that are known to not yet been implemented and won't be implemented via Fortran code so that this patch is an NFC for what is currently supported.

This patch also prevents doing two lookups in the intrinsic table (There was one to get argument lowering rules, and another one to generate the code).

### CODE REVIEW

**Review by tblah — APPROVED** [review]
LGTM, thanks!

File: flang/include/flang/Optimizer/Builder/IntrinsicCall.h [hunk 1]
```diff
@@ -25,6 +25,7 @@
 namespace fir {
 
 class StatementContext;
+struct IntrinsicHandlerEntry;
 
 // TODO: Error handling interface ?
 // TODO: Implementation is incomplete. Many intrinsics to tbd.
```
Comment (vdonaldson) [inline flang/include/flang/Optimizer/Builder/IntrinsicCall.h:31]: You might double check if these comments are still useful and accurate.

**Review by vdonaldson — APPROVED** [review]
Thanks Jean!

File: flang/include/flang/Optimizer/Builder/IntrinsicCall.h (hunk 1 above)
Comment (vdonaldson) [reply flang/include/flang/Optimizer/Builder/IntrinsicCall.h:31]: (The comment about `genIntrinsicCall` looks odd.)

**Review by psteinfeld — APPROVED** [review]
All builds and tests correctly and looks good.

File: flang/include/flang/Optimizer/Builder/IntrinsicCall.h (hunk 1 above)
Comment (jeanPerier) [reply flang/include/flang/Optimizer/Builder/IntrinsicCall.h]: Thanks Val, these TODO are not very helpful/accurate here anymore. I removed them. I also rephrased the `genIntrinsicCall` comment that pointed to another version "above" that does not exists anymore.

**Pull request merged**